core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -311,16 +313,46 @@ class BlockManagerMasterEndpoint(
   }
 
   private def removeShuffle(shuffleId: Int): Future[Seq[Boolean]] = {
-    // Nothing to do in the BlockManagerMasterEndpoint data structures
+    // Find all shuffle blocks on executors that are no longer running
+    val blocksToDeleteByShuffleService =
+      new mutable.HashMap[BlockManagerId, mutable.HashSet[BlockId]]
+    mapOutputTracker.shuffleStatuses.get(shuffleId).map { shuffleStatus =>
+      shuffleStatus.mapStatuses
+        .filter(_.location.port == externalShuffleServicePort)
+        .foreach { mapStatus =>
+          val blocks = blocksToDeleteByShuffleService.getOrElseUpdate(mapStatus.location,
+            new mutable.HashSet[BlockId])
+          val blocksToDel =
+            shuffleManager.shuffleBlockResolver.getBlocksForShuffle(shuffleId, mapStatus.mapId)
+          blocksToDel.foreach(blocks.add(_))
+      }
+    }
+
     val removeMsg = RemoveShuffle(shuffleId)
-    Future.sequence(
-      blockManagerInfo.values.map { bm =>
-        bm.storageEndpoint.ask[Boolean](removeMsg).recover {
-          // use false as default value means no shuffle data were removed
-          handleBlockRemovalFailure("shuffle", shuffleId.toString, bm.blockManagerId, false)
+    val removeShuffleFromExecutorsFutures = blockManagerInfo.values.map { bm =>
+      bm.storageEndpoint.ask[Boolean](removeMsg).recover {
+        // use false as default value means no shuffle data were removed
+        handleBlockRemovalFailure("shuffle", shuffleId.toString, bm.blockManagerId, false)
+      }
+    }.toSeq
+
+    val removeShuffleBlockViaExtShuffleServiceFutures =
+      externalBlockStoreClient.map { shuffleClient =>
+        blocksToDeleteByShuffleService.map { case (bmId, blockIds) =>
```

Comment:
Does `blocksToDeleteByShuffleService` contain blocks of live executors as well?

Reply:
Hmmm it's not supposed to but looking through the code now I didn't seem to add that check in. The RDD removal does it based on whether the block manager ID's port matches the shuffle service port or not. But for shuffles I think that will always be the external shuffle service if it's enabled, so I just need to add a filter for executors that don't exist anymore

Reply:
Add a check to ignore things where the executor is still alive